File: bot.py
import numpy as np
from PIL import Image
import pandas as pd

from pynput.mouse import Controller, Button
import random
import undetected_chromedriver as uc
from selenium import webdriver
from selenium_stealth import stealth
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
import time

# Path to the external text file containing account credentials
credentials_file = "credentials.txt"

# Specify the file path of the Excel file
excel_file = 'AKUN GMAIL.xlsx'

proxy_file_path = 'proxy.txt'

# Read Proxies texts from the text file
proxies = []
with open(proxy_file_path, 'r') as file:
    for line in file:
        proxies.append(line.strip())
# Instantiate the Chrome WebDriver
# driver = webdriver.Chrome()
# Add any desired options to the options object, if needed
# options = uc.ChromeOptions()
# driver = uc.Chrome(options=options)

options = webdriver.ChromeOptions()
options.add_argument("start-maximized")

# Add the --mute-audio flag
options.add_argument("--mute-audio")

# options.add_argument(f'--proxy-server={random.choice(proxies)}')

# options.add_argument("--headless")

# URL to navigate to after logging in
target_url = "https://www.tiktok.com/@nikxphreaker/video/6993669592551836930"

# Path to the external text file containing comment texts
comments_file = "comments.txt"

# Read account credentials from the excel file
accounts = []

# Specify the sheet name (if it's not the default sheet)
sheet_name = 'MARSHA'

# Specify the columns and rows you want to extract
columns = ['Gmail', 'Password.2']  # Replace with your desired column names
start_row = 85  # Replace with the starting row index
end_row = 91  # Replace with the ending row index

# Read the Excel file into a DataFrame
df = pd.read_excel(excel_file, sheet_name=sheet_name)

selected_data = df.loc[start_row:end_row, [columns[0], columns[1]]]
data = pd.DataFrame(selected_data)

# Drop rows with NaN values
data = data.dropna()

# Convert the DataFrame to a list of dictionaries
result = data.to_dict(orient='records')

# Separate the values per row with a comma
for item in result:
    # Skip even numbers
    if item[columns[0]] == str('nan') or item[columns[1]] == str('nan') :
        continue
    account = {"username": item[columns[0]], "password": item[columns[1]]}
    accounts.append(account)

# Read comment texts from the text file
comments = []
with open(comments_file, "r") as file:
    for line in file:
        comment = line.strip()
        comments.append(comment)

# Iterate over each account
for account in accounts:
    
    try:
        # Set the proxy IP address and port
        proxy = random.choice(proxies).split(":")
        proxy_ip = proxy[0] #'199.102.105.242'
        proxy_port = proxy[1] #'4145'

        # Configure the Selenium WebDriver to use the proxy server
        proxy_options = {
            'proxy': {
                'proxyType': 'MANUAL',
                'httpProxy': f'{proxy_ip}:{proxy_port}',
                'sslProxy': f'{proxy_ip}:{proxy_port}'
            }
        }

        # options.add_argument('user_agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.53 Safari/537.36')
        options.add_experimental_option("excludeSwitches", ["enable-automation"])
        options.add_experimental_option('useAutomationExtension', False)
        driver = webdriver.Chrome(options=options)

        stealth(driver,
                languages=["en-US", "en"],
                user_agent= 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.53 Safari/537.36',
                vendor="Google Inc.",
                platform="Win32",
                webgl_vendor="Intel Inc.",
                renderer="Intel Iris OpenGL Engine",
                fix_hairline=True,
                proxy_options=proxy_options
                )
        print(f'Account {account["username"]} is logging in...')

        # Open TikTok website
        driver.get("https://www.tiktok.com/login/phone-or-email/email")

        # Wait for the page to load
        time.sleep(2)

        # Find the login button and click it
        # login_button = driver.find_element(By.LINK_TEXT, "Log in")
        # login_button.click()

        # Wait for the login page to load
        time.sleep(2)

        # Find the username input field and enter the username
        username_input = driver.find_element(By.CSS_SELECTOR, "input[name='username']")
        username_input.send_keys(account["username"])
        time.sleep(5)

        # Find the password input field and enter the password
        password_input = driver.find_element(By.CSS_SELECTOR, "input[type='password']")
        password_input.send_keys(account["password"])
        time.sleep(5)

        # Press Enter to submit the login form
        password_input.send_keys(Keys.ENTER)

        # ############################################
        # #### Resolve captcha
        # ############################################

        # Wait for the login process to complete
        time.sleep(15)

        # Navigate to the target URL
        driver.get(target_url)

        time.sleep(5)


        ############################################
        #### Bot for like and comment
        ############################################
        # Add likes to the post
        wait = WebDriverWait(driver, 10)
        like_button = wait.until(EC.element_to_be_clickable((By.XPATH, '/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]')))
        like_button.click()
        time.sleep(5)

        # Add comment to the post
        # Select a random comment from the list
        comment_text = random.choice(comments)
        # Create an instance of ActionChains
        # Move to the element to make it visible
        comment_click = driver.find_element(By.XPATH, '/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div')
        actions = ActionChains(driver)
        actions.move_to_element(comment_click)
        actions.perform()
        comment_click.click()
        comment_input = wait.until(EC.presence_of_element_located((By.XPATH, '/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div/div[1]/div/div/div[2]/div/div/div/div/span')))
        comment_input.send_keys(comment_text)
        time.sleep(2)
        comment_input.send_keys(Keys.ENTER)
        time.sleep(2)

        # Find a video element and extract the video URL
        #video_element = driver.find_element(By.CSS_SELECTOR, "a[href^='/v/']")
        #video_url = video_element.get_attribute("href")
        #print("Video URL:", video_url)

        # Move to the next video
        # driver.find_element_by_tag_name("body").send_keys(Keys.ARROW_RIGHT)
        
        # Logout and clear cookies for the next account
        driver.delete_all_cookies()
        print(f'Account {account["username"]} Success...')
    except ValueError:
        print(f'Account {account["username"]} Failed...')
        # Code to handle ValueError exception
        print("ValueError occurred!")
    except TypeError:
        print(f'Account {account["username"]} Failed...')
        # Code to handle TypeError exception
        print("TypeError occurred!")
    except Exception as e:
        print(f'Account {account["username"]} Failed...')
        # Code to handle any other exceptions


# Close the browser
driver.quit()

File: bot.js
const webdriver = require('selenium-webdriver');
const { Builder, By, Key, until } = webdriver;
const chrome = require('selenium-webdriver/chrome');
const CaptchaSolver = require('tiktok-captcha-solver')

const logged_in_accounts = new Set(); // Keep track of logged-in accounts

const ua = require('random-useragent');
const user_agent = ua.getRandom();

async function loginTikTok(accounts, proxies, targetUrl, comments) {
  // Iterate over each account
  for (const account of accounts) {
    const options = new chrome.Options();
    options.addArguments('start-maximized');
    options.addArguments('--mute-audio');
    options.addArguments('--no-sandbox');
    options.addArguments('--disable-dev-shm-usage');
    options.addArguments(`--user-agent=${user_agent}`);
    options.excludeSwitches('enable-automation');
    options.setUserPreferences({ 'useAutomationExtension': false });

    const chromePath = 'chromedriver.exe';
    
    // const service = new chrome.ServiceBuilder().build();
    // const service = new chrome.ServiceBuilder(chromePath).build();
    // console.log(chrome);
    // chrome.setDefaultService(service);

    const driver = new Builder()
      .forBrowser('chrome')
      .setChromeOptions(options)
      .build();

    // const wait = new webdriver.WebDriverWait(driver, 10);

    const username = account.username;
    const password = account.password;
    const proxy = proxies[Math.floor(Math.random() * proxies.length)];
    const comment = comments[Math.floor(Math.random() * comments.length)];

    if (logged_in_accounts.has(username)) {
      console.log(`Account ${username} is already logged in. Skipping...`);
      return;
    }

    // // Set up Chrome options
    // const chromeOptions = {
    //   'args': [
    //     '--mute-audio',
    //     `--proxy-server=${proxy}`,
    //     '--no-sandbox',
    //     '--disable-dev-shm-usage',
    //     `--user-agent=${user_agent}`,
    //   ]
    // };

    // // Create a new WebDriver instance
    // const driver = await new Builder()
    //   .forBrowser('chrome')
    //   .setChromeOptions(chromeOptions)
    //   .build();

    try {
      console.log(`Account ${username} is logging in...`);

      // the captcha solver should be initialized before visiting the page
      // const captchaSolver = new CaptchaSolver(page)

      // await page.goto('https://www.tiktok.com/@tiktok')

      // // solve the captcha
      // await captchaSolver.solve()

      await driver.get('https://www.tiktok.com/login/phone-or-email/email');
      await driver.wait(until.elementLocated(By.css('input[name="username"]')), 10000);

      const emailField = await driver.findElement(By.css('input[name="username"]'));
      await emailField.sendKeys(username);

      await driver.wait(until.elementLocated(By.css('input[type="password"]')), 10000);
      const passwordField = await driver.findElement(By.css('input[type="password"]'));
      await passwordField.sendKeys(password);
      await driver.sleep(5000);

      // Press Enter to submit the login form
      await passwordField.sendKeys(Key.ENTER);

      await driver.wait(until.urlIs('https://www.tiktok.com'), 10000);

      // Example: Check if login was successful
      if (driver.currentUrl.includes('login')) {
        throw new Error('Login failed');
      } else {
        // Navigate to the target URL
        await driver.get(targetUrl);
      }

      // Add the account to the set of logged-in accounts
      logged_in_accounts.add(username);

      // Bot for like and comment
      const likeButton = await driver.wait(until.elementToBeClickable(By.xpath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]')));
      const tagElement = await driver.wait(until.elementToBeClickable(By.xpath('/html/body/div[1]/div[3]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]/span/div/div/svg/g/g[2]/g/g[4]')));

      // Get the value of the 'style' attribute
      const styleValue = await tagElement.getAttribute('style');

      // Perform a conditional action based on the style value
      if (styleValue.includes('display: none;')) {
        // Do something if the style value contains 'display: none;'
        await likeButton.click();
        await driver.sleep(5000);

        // Add comment to the post
        // Create an instance of ActionSequence
        // Move to the element to make it visible
        const commentClick = await driver.findElement(By.xpath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div'));
        const actions = driver.actions({ bridge: true });
        await actions.move({ origin: commentClick }).perform();
        await commentClick.click();

        const commentInput = await driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div/div[1]/div/div/div[2]/div/div/div/div/span')));
        await commentInput.sendKeys(comment);
        await driver.sleep(2000);
        await commentInput.sendKeys(Key.ENTER);
        await driver.sleep(2000);
      }

      // Logout and clear cookies for the next account
      await driver.manage().deleteAllCookies();
    } catch (error) {
      console.log(`Error during login: ${error.message}`);
    } finally {
      await driver.quit();
    }
  }
}

async function main() {
  const workers = 5; // multi-instance

  // Specify the file path of the Excel file
  const excelFile = 'AKUN GMAIL.xlsx';

  // Path to the external text file containing comment texts
  const commentsFile = 'comments.txt';

  // Path to the text file containing proxy addresses
  const proxyFilePath = 'proxy.txt';

  // URL to navigate to after logging in
  const targetUrl = 'https://www.tiktok.com/@nikxphreaker/video/6993669592551836930';

  // Read account credentials from the Excel file
  const accounts = [];
  const sheetName = 'MARSHA'; // Specify the sheet name if it's not the default sheet

  // Specify the columns and rows you want to extract
  const columns = ['Gmail', 'Password.2']; // Replace with your desired column names
  const startRow = 50; // Replace with the starting row index
  const endRow = 54; // Replace with the ending row index

  // Read the Excel file
  const XLSX = require('xlsx');
  const workbook = XLSX.readFile(excelFile);
  const worksheet = workbook.Sheets[sheetName];

  const data = XLSX.utils.sheet_to_json(worksheet, { header: 1 });

  // Extract the desired data range
  const selectedData = data.slice(startRow - 1, endRow);

  // Find the column indices based on column names
  // const column1Index = selectedData.indexOf(columns[0]);
  // const column2Index = selectedData.indexOf(columns[1]);

  // Extract data from the specified columns
  // const dataArray = selectedData.map(row => {
  //   if (row[1] != undefined || row[10] != undefined) {
  //     continue; // Skip rows with invalid data
  //   } else {
  //     [row[1], row[10]]
  //   }
    
  // });

  // console.log(selectedData);
  // console.log(dataArray);

  // // Convert the selected data into a list of objects
  for (const row of selectedData) {
    const username = row[1];
    const password = row[10];

    if (username == undefined || password == undefined) {
      continue; // Skip rows with invalid data
    }

    const account = { 'username' : username, 'password' : password };
    console.log(account);
    accounts.push(account);
  }

  // Read proxy addresses from the text file
  const fs = require('fs');
  const proxies = fs.readFileSync(proxyFilePath, 'utf-8').split('\n').map(line => line.trim()).filter(line => line !== '');

  // Read comment texts from the text file
  const comments = fs.readFileSync(commentsFile, 'utf-8').split('\n').map(line => line.trim()).filter(line => line !== '');

  const numPieces = 5; // Number of pieces to split the array into
  const arrayLength = accounts.length;
  const pieceSize = Math.floor(arrayLength / numPieces);
  const remainder = arrayLength % numPieces;

  const splitArrays = [];
  let start = 0;
  for (let i = 0; i < numPieces; i++) {
    const end = start + pieceSize + (i < remainder ? 1 : 0); // Adjust piece size for the remainder
    splitArrays.push(accounts.slice(start, end));
    start = end;
  }

  //Create an array of promises for each WebDriver instance
  const promises = splitArrays.map((accounts) => loginTikTok(accounts, proxies, targetUrl, comments));

  // Execute all promises concurrently
  await Promise.all(promises);

  console.log('All Selenium instances have completed.');

  // try {
  //   for (const account of splitArrays) {
  //     const username = account.username;
  //     const password = account.password;
  //     const proxy = proxies[Math.floor(Math.random() * proxies.length)];
  //     const comment = comments[Math.floor(Math.random() * comments.length)];

  //     if (logged_in_accounts.has(username)) {
  //       console.log(`Account ${username} is already logged in. Skipping...`);
  //       return;
  //     }

  //     console.log(`Account ${username} is logging in.`);

  //     await driver.get('https://www.tiktok.com/login/phone-or-email/email');

  //     await driver.wait(until.elementLocated(By.css('input[name="username"]')));

  //     const emailField = await driver.findElement(By.css('input[name="username"]'));
  //     await emailField.sendKeys(username);

  //     await driver.wait(until.elementLocated(By.css('input[type="password"]')));

  //     const passwordField = await driver.findElement(By.css('input[type="password"]'));
  //     await passwordField.sendKeys(password);
  //     await driver.sleep(5000);

  //     // Press Enter to submit the login form
  //     await passwordField.sendKeys(Key.ENTER);

  //     await driver.wait(until.urlIs('https://www.tiktok.com'));

  //     // Example: Check if login was successful
  //     if (driver.getCurrentUrl().includes('login')) {
  //       throw new Error('Login failed');
  //     } else {
  //       // Navigate to the target URL
  //       await driver.get(targetUrl);
  //     }

  //     // Add the account to the set of logged-in accounts
  //     logged_in_accounts.add(username);

  //     // Bot for like and comment
  //     const likeButton = await driver.wait(until.elementToBeClickable(By.xpath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]')));
  //     const tagElement = await driver.wait(until.elementToBeClickable(By.xpath('/html/body/div[1]/div[3]/div[2]/div/div[2]/div[1]/div[1]/div[1]/div[4]/div/button[1]/span/div/div/svg/g/g[2]/g/g[4]')));

  //     // Get the value of the 'style' attribute
  //     const styleValue = await tagElement.getAttribute('style');

  //     // Perform a conditional action based on the style value
  //     if (!styleValue.includes('display: none;')) {
  //       await likeButton.click();
  //       await driver.sleep(5000);

  //       // Add comment to the post
  //       const commentClick = await driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div')));
  //       await driver.actions({ bridge: true }).move({ origin: commentClick }).perform();
  //       await commentClick.click();

  //       const commentInput = await driver.wait(until.elementLocated(By.xpath('/html/body/div[1]/div[2]/div[2]/div/div[2]/div[1]/div[3]/div[1]/div/div/div[1]/div/div[1]/div/div/div[2]/div/div/div/div/span')));
  //       await commentInput.sendKeys(comment);
  //       await driver.sleep(2000);
  //       await commentInput.sendKeys(Key.ENTER);
  //       await driver.sleep(2000);
  //     }

  //     // Logout and clear cookies for the next account
  //     await driver.manage().deleteAllCookies();
  //   }
  // } catch (error) {
  //   console.log(`Error during login: ${error.message}`);
  // } finally {
  //   await driver.quit();
  // }
}

main().catch(console.error);
